LIME-92: Add stripe integration

ArtemLymarenko opened the pull request:
This PR contains general implementation of subscruption logic. All features that were implemented showed on the video.
The actual flow of subscription looks like this:
1. When user try to subscribe, automaticly in subscription table  created row with status incomplete. If payment success via webhook i change status to active.
2. If payment remains incomlete, after 24 hours stripe automaticly mark it as cancelled and through webhook I delete it from subscription table.
3. If user wants to cancel his subscription he can do a soft cancel, this means that subscription will be totaly cancelled and deleted from table at the end of billing period. User also can cancel this action by renewing a subscription.
4. If user changed his subscription. Previous subscription cancels and deletes from table and new subscription become availiable after successfull payment.

Any comments or suggestions to improve this implementation would be greatly appreciated.

**You can see in the video that the expiration date of the next subscription is 27.05. This is because I previously tested simulation time feature, so the current date in stipe for this user is 27.04.**

https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/92994963/73075456-e798-4a58-9d70-fd25869a083e

dimapopovych commented on frontend/src/bundles/subscription/pages/subscription-checkout/subscription-checkout.tsx:
```diff
@@ -0,0 +1,39 @@
+const stripe = loadStripe(config.ENV.STRIPE.PUBLIC_KEY);
+
+const SubscriptionCheckout = (): JSX.Element => {
+    const { clientSecret } = useAppSelector(({ subscriptions }) => {
+        return {
+            clientSecret: subscriptions?.clientSecret as string,
+        };
+    });
```
why do you need to store client secret in store when you can retrieve it from config?

ArtemLymarenko replied:
As I understood the docs, first I have to retrieve it from subscribe request when creating a new subscription and use it to complete payment. Only after successful payment I can retrieve it from url, to display some messages about results of payment operation.

ArtemLymarenko replied:
![изображение](https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/92994963/b8e4e3e0-6bc7-4a69-8695-b1799f683cca)

dimapopovych replied:
My bad, I forgot that the intent also returns the client_secret and mistakenly thought you were storing the application's client secret in store for some reason.

dimapopovych commented on backend/src/bundles/subscription-plans/subscription-plan.entity.ts:
```diff
@@ -0,0 +1,120 @@
+
+    private 'description': string | null;
+
+    private 'productToken': string;
+
+    private 'priceToken': string;
```
is it token? Wouldn't be better to name it stripe_product_id etc.?

ArtemLymarenko replied:
Should I name stripe_token as stripe_subscription_id too? Because for me it was a bit confusing to name it like subscription_id, price_id etc.

dimapopovych replied:
Yes, I meant using similar approach for everything. As for me using the prefix for such fields clearer shows that they are third-party identifiers.

dimapopovych commented on backend/src/bundles/subscriptions/enums/enums.ts:
```diff
@@ -0,0 +1,9 @@
+export { SubscriptionWebHook } from './subscription-webhooks-enum.js';
+export {
+    HttpCode,
+    HttpError,
```
import from common/http

dimapopovych commented on backend/src/bundles/subscriptions/enums/subscription-attributes.enum.ts:
```diff
@@ -0,0 +1,13 @@
+    USER_ID: 'userId',
+    STATUS: 'status',
+    SUBSCRIPTION_TOKEN: 'subscriptionToken',
+    CANCEL_AT_PERIOD_END: 'cancelAtPeriodEnd',
+    EXPIRATION_DATE: 'expirationDate',
```
let's change namings to isCanceled and expiresAt

SofiiaTrokhymchuk commented on backend/src/bundles/subscription-plans/subscription-plan.service.ts:
```diff
@@ -0,0 +1,21 @@
+import { type SubscriptionPlansGetAllResponseDto } from './types/types.js';
+
+class SubscriptionPlanService
+    implements Omit<Service, 'find' | 'create' | 'update' | 'delete'>
```
You can just avoid implementing if you are sure that methods from the interface are not needed

nikita-remeslov commented:
Is this PR ready?

ArtemLymarenko commented:
I need to rename variables as required and add docs for swagger. Also I'll merge branches to resolve conflicts. I created it as draft to check if my logic of implementation is fine.

ArtemLymarenko commented:
Changed style of cancel button and added this "danger" style to common button component.

![изображение](https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/92994963/2637247b-36d2-4de4-bc3d-0e690814a6cb)

dimapopovych commented on backend/src/bundles/subscriptions/subscription.service.ts:
```diff
@@ -0,0 +1,182 @@
+                        userId,
+                    );
+
+                if (activeSubscriptions && activeSubscriptions.length > 1) {
+                    for (const [
+                        index,
+                        subscription,
+                    ] of activeSubscriptions.entries()) {
+                        if (index !== 0) {
+                            const { stripeSubscriptionId } =
+                                subscription.toNewObject();
+                            await stripeService.immediateCancelSubscription({
+                                stripeSubscriptionId,
+                            });
+                        }
+                    }
+                }
```
why there is possibility that user will have more than one active subscription?

ArtemLymarenko replied:
When user changes his subscription, I create new subscription in stripe and the old subscription cancels.

anton-otroshchenko commented on backend/src/common/helpers/format-to-date-from-unix-helper.ts:
```diff
@@ -0,0 +1,5 @@
+const formatToDateFromUnix = (unixDate: number): Date => {
+    return new Date(unixDate * 1000);
```
Create a constant for 1000 because its unclear what does it mean

anton-otroshchenko commented on backend/src/common/services/stripe/stripe.service.ts:
Create types for each object in this file to follow the project structure

dimapopovych commented on backend/src/bundles/subscriptions/subscription.model.ts:
```diff
@@ -0,0 +1,57 @@
+    public static override get relationMappings(): RelationMappings {
+        return {
+            subscriptionPlan: {
+                relation: Model.HasOneRelation,
```
this should be BelongsToOne, since you have FK for this relation in this model

dimapopovych commented on backend/src/bundles/subscriptions/subscription.model.ts:
```diff
@@ -0,0 +1,57 @@
+                },
+            },
+            user: {
+                relation: Model.HasOneRelation,
```
same

dimapopovych commented on backend/src/bundles/subscriptions/subscription.repository.ts:
```diff
@@ -0,0 +1,124 @@
+    ): Promise<SubscriptionEntity[] | null> {
+        const subscriptions = await this.subscriptionModel
+            .query()
+            .where(SubscriptionAttributes.USER_ID, userId)
+            .andWhere(SubscriptionAttributes.STATUS, SubscriptionStatus.ACTIVE)
```
could just combine into one where clause

dimapopovych commented on backend/src/bundles/subscriptions/subscription.repository.ts:
```diff
@@ -0,0 +1,124 @@
+            .query()
+            .findOne(query)
+            .where(SubscriptionAttributes.STATUS, SubscriptionStatus.ACTIVE)
+            .orderBy(SubscriptionAttributes.CREATED_AT, 'DESC')
```
why do you need orderBy here? You have only one row that will be returned.

ArtemLymarenko replied:
Sorry, forgot to change

dimapopovych commented on backend/src/bundles/subscriptions/subscription.repository.ts:
```diff
@@ -0,0 +1,124 @@
+    ): Promise<SubscriptionEntity> {
+        const subscription = await this.subscriptionModel
+            .query()
+            .insert({
+                ...entity.toNewObject(),
+            })
```
seems like unnecessary spread

dimapopovych commented on backend/src/bundles/users/user.model.ts:
```diff
@@ -40,6 +46,14 @@ class UserModel extends AbstractModel {
                 },
             },
+            subscriptions: {
+                relation: Model.HasOneRelation,
```
HasMany? you said that user could potentially have multiple subscription records in DB

ArtemLymarenko replied:
Can I remove this relation since I'm not using it?

dimapopovych replied:
If you remove this relation from the model, it won't magically disappear, it will still exist but just won't be declared) Someone reading the code later might have a hard time determining the presence of this relation. So, I would say keep it to ensure our models closely represent the actual state of the database, even if you don't use it now.

ArtemLymarenko replied:
Ok. You're right!

dimapopovych commented on backend/src/common/server-application/base-server-app.ts:
```diff
@@ -126,6 +128,22 @@ class BaseServerApp implements ServerApp {
+            runFirst: true,
+            routes: this.apis.map((it) => {
+                return it.buildFullPath(
+                    ApiPath.SUBSCRIPTIONS + SubscriptionsApiPath.STRIPE_WEBHOOK,
```
let's use strings interpolation here

dimapopovych commented on frontend/src/bundles/subscription/components/subscription-plan/subscription-plan.tsx:
```diff
@@ -0,0 +1,59 @@
+import { useCallback } from 'react';
```
import

dimapopovych commented on frontend/src/bundles/subscription/components/subscription-plan/subscription-plan.tsx:
```diff
@@ -0,0 +1,59 @@
+    description: string;
+    price: number;
+    stripePriceId: string;
+    handleClick: ({ planId, stripePriceId }: SubscribeRequestDto) => void;
```
should be prefixed with on, quality criteria A9
This is likely the first time I've seen this type of type declaration syntax used in code)

dimapopovych commented on frontend/src/bundles/subscription/components/subscription-user-panel/subscription-user-panel.tsx:
```diff
@@ -0,0 +1,107 @@
+    isCanceled: boolean;
+    expiresAt: Date;
+    stripeSubscriptionId: string;
+    handleChangeSubscription: () => void;
```
same

nikita-remeslov reviewed (approved):
You really exceeded the task) it’s better to fix this next time.
But overall very cool work!
